js/noms/src/value-store.js

```diff
@@ -188,47 +192,51 @@ class SizeCache<T> {
     return entry;
   }
 
-  get(hash: Hash): ?T {
-    const entry = this.entry(hash);
-    return entry ? entry.value : undefined;
-  }
-
-  add(hash: Hash, size: number, value: ?T): void {
+  add(hash: Hash, size: Promise<number>, value: T): void {
     const key = hash.toString();
-    if (this._cache.has(key)) {
+    let entry = this._cache.get(key);
+    if (entry) {
       this._cache.delete(key);
-    } else {
-      this._size += size;
+      entry.size.then(() => this.expire());
     }
-    this._cache.set(key, new CacheEntry(size, value));
 
-    if (this._size > this._maxSize) {
-      for (const [key, {size}] of this._cache) {
-        if (this._size <= this._maxSize) {
-          break;
-        }
-        this._cache.delete(key);
-        this._size -= size;
-      }
+    entry = new CacheEntry(size, value);
+    this._cache.set(key, entry);
+    entry.size.then(() => this.expire());
```

Comment:
I don't understand this line... it means you might call `this.expire()` twice on cache entries, each time they're added, then potentially a 2nd time if they're replaced? I guess it's idempotent, but it seems unnecessary?